Feat/cgroup v2 support (#276)

DhanashreePetare opened the pull request:
## Cgroup v1/v2 Monitoring Support

### Overview
Adds comprehensive cgroup (container) resource monitoring to prmon with automatic detection of cgroup v1, v2, and hybrid environments.

### Changes
- **New monitor**: `cgroupmon` with 17 metrics across CPU, memory, I/O, and process counts
- **Auto-detection**: Detects cgroup version at runtime and adapts accordingly
- **Parser**: Handles `cpu.stat`, `memory.stat`, `io.stat` for both v1 and v2
- **Tests**: Python integration test + GTest skeleton with precooked fixtures
- **Docs**: README section, ADDING_MONITORS example, CONTRIBUTING Windows guide
- **Disable flag**: `--disable cgroupmon` if needed

### Benefits
- Better container/Kubernetes workload monitoring
- More accurate I/O stats in containerized environments
- Backward compatible with existing prmon deployments
- Graceful degradation if cgroups unavailable

### Testing
Python linting passes. CI will verify build and tests in Linux environments.

Commit:
feat: Add comprehensive cgroup v1/v2 monitoring support

- Implement cgroupmon monitor with 17 metrics across CPU, memory, I/O, and process counts
- Support automatic detection of cgroup v1, v2, and hybrid environments
- Parse cgroup controllers: cpu.stat, memory.stat, io.stat for both versions
- Integrate with prmon build system and test infrastructure
- Add Python and GTest test harnesses with precooked cgroup v2 fixtures
- Update documentation: README, ADDING_MONITORS, CONTRIBUTING with usage and Windows dev notes
- Enable container resource tracking with --disable cgroupmon flag
- Provide more accurate I/O statistics for containerized workloads

Commit:
fix: cgroupmon logic errors - remove unused variable and fix proc/thread counting

- Remove unused cgroup_stat_update variable in update_stats()
- Fix read_single_value() to count lines for cgroup.procs and cgroup.threads
  instead of reading first value, since these are multi-line PID lists

graeme-a-stewart reviewed (changes requested):
See comment above for restructuring changes.

graeme-a-stewart commented:
Hello @DhanashreePetare 

Is this an issue you would still like to work on? Please let us know and we can try to assign to another person if you are not able to continue.

Thanks.

DhanashreePetare commented:
sorry @graeme-a-stewart, very sorry for my late reply. Yes I will change the workflow according to the above provided changes, I will get back very soon with the changes.

Commit:
refactor: Implement maintainer feedback - v2-only cgroup support with focused metrics

Based on maintainer guidance:
- Remove cgroup v1 and hybrid support (v2-only going forward)
- Remove 3 less useful metrics: cgroup_mem_slab, cgroup_mem_pgfault, cgroup_mem_pgmajfault
- Simplify logic for easier maintenance (14 metrics, no v1/v2 branching)
- Fix get_hardware_info() to avoid referencing removed enum values
- Update all documentation to reflect v2-only implementation

Changes:
- Enum: NONE, V1, V2, HYBRID → NONE, V2
- Metrics: 17 → 14 (removed: slab, pgfault, pgmajfault)
- Removed functions: read_cgroup_v1_stats, parse_*_v1 (4 functions)
- Simplified detection, path resolution, and hardware info logic
- Updated tests, README, ADDING_MONITORS, CGROUP_IMPLEMENTATION docs

graeme-a-stewart commented on doc/CONTRIBUTING.md:
```diff
@@ -65,6 +65,46 @@ issue to first discuss with the maintainers.
+- Use WSL2 (Ubuntu) or Docker Desktop to provide a Linux runtime.
+- Run all builds and tests inside WSL2 or a Linux container.
+
+For formatting on Windows before opening a pull request:
```
Windows is not a supported environment for prmon, so while I appreciate your instructions, it's just adding noise. If one is using Windows or OS X then a container / WSL2 environment needs to be used and then you have your Linux for build, tests, formatting, etc.

graeme-a-stewart commented on doc/CONTRIBUTING.md:
````diff
@@ -65,6 +65,46 @@ issue to first discuss with the maintainers.
+
+```powershell
+cmake -S . -B build
+cmake --build build --target clang-format
````
This is already documented (L59).

graeme-a-stewart commented on package/src/cgroupmon.cpp:
```diff
@@ -0,0 +1,291 @@
+// Copyright (C) 2018-2025 CERN
```
This is a new file, so it should be (C) for the dates the file existed, viz, 2025-2026 in this case.

graeme-a-stewart commented on package/src/cgroupmon.cpp:
```diff
@@ -0,0 +1,291 @@
+  
+  std::string line;
+  // For cgroup v2, look for line starting with "0::"
+  // For cgroup v2, look for line starting with "0::"
```
Duplicated comment.

graeme-a-stewart commented on package/src/cgroupmon.cpp:
```diff
@@ -0,0 +1,291 @@
+// Find the cgroup filesystem mount point
+std::string cgroupmon::find_cgroup_mount_point() {
+  return "/sys/fs/cgroup";
+}
```
This is now OTT for returning the same string every time (and you're not "detecting" anything). Instead just define a `const std::string`.

graeme-a-stewart commented on package/src/cgroupmon.cpp:
```diff
@@ -0,0 +1,291 @@
+  }
+
+  // Detect cgroup version
+  cgroup_version = detect_cgroup_version();
```
This now seems odd - we only support one version (v2) and we just detect if this seems to be supported. `CgroupVersion::NONE` isn't really a version at all.

Can you change the function naming and logic here to then be a bit clearer? e.g., you can have `detect_cgroup_v2()` returning a boolean instead. I think this will work and it will be clearer.

graeme-a-stewart commented on package/src/cgroupmon.cpp:
```diff
@@ -0,0 +1,291 @@
+  }
+
+  // Find cgroup mount point
+  cgroup_mount_point = find_cgroup_mount_point();
```
```cpp
const std::string cgroup_mount_point = "/sys/fs/cgroup";
```

would be clearer and avoid a useless function.

graeme-a-stewart commented on package/src/cgroupmon.cpp:
```diff
@@ -0,0 +1,291 @@
+
+  // Find cgroup mount point
+  cgroup_mount_point = find_cgroup_mount_point();
+  if (cgroup_mount_point.empty()) {
```
Right now this does nothing at all, as `cgroup_mount_point` is always defined. A test that the mount point exists has already been done in `detect_cgroup_version()`.

graeme-a-stewart commented on package/src/cgroupmon.cpp:
```diff
@@ -0,0 +1,291 @@
+
+  // Use first PID to determine cgroup path
+  // (all processes in the tree should be in the same cgroup)
+  std::string base_path = read_path.empty() ? "" : read_path;
```
`base_path` serves no purpose - it's always the same as `read_path`. And the ternary operation is pointless.

graeme-a-stewart commented on package/src/cgroupmon.cpp:
```diff
@@ -0,0 +1,291 @@
+  // Memory current and max
+  stats["cgroup_mem_current"] = read_single_value(cgroup_path + "/memory.current") / 1024;
+  unsigned long long mem_max = read_single_value(cgroup_path + "/memory.max");
+  if (mem_max != ULLONG_MAX) {
```
First, better to use the modern C++ limit here, viz. `std::numeric_limits<unsigned long long>::max()`.

Second, what is the purpose of this test?

graeme-a-stewart commented on package/src/cgroupmon.cpp:
```diff
@@ -0,0 +1,291 @@
+  
+  // Memory current and max
+  stats["cgroup_mem_current"] = read_single_value(cgroup_path + "/memory.current") / 1024;
+  unsigned long long mem_max = read_single_value(cgroup_path + "/memory.max");
```
What happens when `memory.max` has no limit and contains the entry `max`? As far as I can see you would here return `0`, correct?

On reflection I think it is probably the most useful thing to map a `max` value to zero, but this needs to be well documented (it can happen for some other cgroupsv2 values as well).

graeme-a-stewart commented on package/src/cgroupmon.cpp:
```diff
@@ -0,0 +1,291 @@
+  }
+
+  // For single-value files, read the first number
+  unsigned long long value;
```
I think the code here would be better if you mapped the one expected string, `max`, to the value `0` explicitly.

At the moment this is happening silently and implicitly, which is not good.

graeme-a-stewart commented on package/src/cgroupmon.cpp:
```diff
@@ -0,0 +1,291 @@
+    return;
+  }
+  
+  hw_json["cgroup"]["version"] = "v2";
```
This is not useful anymore as cgroups implies v2, and the mount point is fixed. Please remove.

graeme-a-stewart commented on package/src/cgroupmon.h:
```diff
@@ -0,0 +1,95 @@
+#include "registry.h"
+
+// Cgroup version enumeration
+enum class CgroupVersion { NONE, V2 };
```
I think this is no longer needed - see comments above w.r.t. detecting cgroupsv2. An enum is overdesign.

graeme-a-stewart commented on package/src/cgroupmon.h:
```diff
@@ -0,0 +1,95 @@
+  prmon::monitored_list cgroup_stats;
+
+  // Cgroup detection state
+  CgroupVersion cgroup_version;
```
Not needed.

graeme-a-stewart commented on package/src/cgroupmon.h:
```diff
@@ -0,0 +1,95 @@
+  CgroupVersion cgroup_version;
+  bool valid;
+  std::string cgroup_path;
+  std::string cgroup_mount_point;
```
This could now just be a `const std::string`.

graeme-a-stewart commented on CGROUP_IMPLEMENTATION.md:
```diff
@@ -0,0 +1,239 @@
+# Cgroup v2 Support Implementation - Summary
```
We really don't need this file. The historical record of the addition of the cgroups monitoring is in the pull request.

graeme-a-stewart reviewed (changes requested):
Hello @DhanashreePetare - thanks a lot for the changes. There are still a few things to improve on with some further simplifications that will make the code even leaner.

graeme-a-stewart commented:
FYI, I just fixed the `CMakeLists.txt` conflict for you

graeme-a-stewart commented:
There is also a test failure, can you investigate it?

```sh
25/28 Testing: testUnits
25/28 Test: testUnits
Command: "/usr/bin/python3" "test_cpu.py" "--units" "--time" "3" "--slack" "0" "--interval" "1"
Directory: /tmp/build/package/tests
"testUnits" start time: Mar 24 17:33 UTC
Output:
----------------------------------------------------------
Will run for 3s using 1 process(es) and 1 thread(s)
[2026-03-24 17:33:21.529] [nvidiamon] [warning] Failed to load libnvidia-ml.so: libnvidia-ml.so.1: cannot open shared object file: No such file or directory
[2026-03-24 17:33:21.529] [nvidiamon] [warning] Both NVML and nvidia-smi failed. NVIDIA monitoring disabled.
F
======================================================================
FAIL: test_run_test_with_params (__main__.setup_configurable_test.<locals>.ConfigurableProcessMonitor)
Actual test runner
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/build/package/tests/test_cpu.py", line 99, in test_run_test_with_params
    self.assertEqual(
AssertionError: 3 != 0 : Wrong number of unit values for 'Avg' - missing parameters are {'cgroup_cpu_periods', 'cgroup_cpu_throttled', 'cgroup_mem_max'}

----------------------------------------------------------------------
Ran 1 test in 3.075s
```

N.B. You can ignore the `nvidiamon` warning - that needs to be lowered in severity
